File: gpt.py
from openai import OpenAI
#import env     # for Jeannie; uncomment everyone else
import sys
import os

#client = OpenAI(api_key=env.api_key)     # for Jeannie; uncomment everyone else
client = OpenAI(api_key=os.environ["OPENAI_API_KEY"])   # comment out everyone

def get_feedback(word, wills_transcription, students_transcription):
    response = client.chat.completions.create(
        model="gpt-4",
        messages=[
            {
                "role": "system",
                "content": (
                    "You are an assistant of Will Styler, a professor of a linguistics class called LIGN 101. "
                    "You will be assisting him in giving feedback on students’ transcriptions of certain words on an assignment. "
                    "You will be provided the word, the correct transcription, and the student’s answer. "
                    "If the student’s answer matches the correct answer, congratulate them. If not, give them constructive feedback and advice on what aspects of the IPA the student can think about to get their transcription closer to the actual answer. "
                    "Under no circumstances should you include the direct answer or any IPA symbols from the correct answer in your message to the student. "
                    "This is against the rules and wouldn’t help the student learn much from the assignment. "
                    "If the student complicates the answer and gives a transcription outside the scope of the class, tell them which part of their transcription is outside the scope of the class. "
                    "For instance, if the student used diacritics in the transcription, explain to them that diacritics are outside the scope of the class and try again without them. "
                    "Remember that you will be talking directly to the student, and limit your response to about 200 words. \n\n"
                    "Things to make sure before you give the feedback:\n"
                    "Do not include any IPA characters whatsoever in your response. "
                    "For instance, do not say something like, “The student's transcription of /gool/ differs from the standard transcription /gul/.”. "
                    "This is against the rules of the class.\n"
                    "You will be talking directly to the students, so format your response such that the students understand that you are talking directly to them. \n"
                    "Highlight the parts of the student’s transcriptions that differs from the correct transcription.\n"
                    "Remember that the correct transcriptions in this class may or may not be the standard transcriptions. "
                    "So don't call our correct answers “standard transcriptions”. This can be misleading to the students. Instead, call them 'Will's transcriptions', as these transcriptions are based on Will Styler's dialect.\n"
                    "Consider the possible reasons of why the student had that answer. If you think it might've been something to do with the student's accent, mention that as a part of your feedback.\n"
                    "Explain why the student's transcription is wrong. If certain parts of the transcription is wrong, for instance, if a student used 'ʒ' instead of 'z', find words which uses these symbols and show the student the difference between the two symbols. \n\n"
                    "DO NOT GIVE AWAY THE CORRECT TRANSCRIPTION\n"
                )
            },
            {
                "role": "user",
                "content": f"WORD: {word}\nWill's transcription: {wills_transcription}\nStudent's transcription: {students_transcription}"
            }
        ],
        temperature=0.25,
        max_tokens=200,
        top_p=1,
        frequency_penalty=0,
        presence_penalty=0
    )
    
    return response.choices[0].message.content

if __name__ == "__main__":
    word = sys.argv[1]
    wills_transcription = sys.argv[2]
    students_transcription = sys.argv[3]
    feedback = get_feedback(word, wills_transcription, students_transcription)
    print(feedback)

File: gpt.js
require('dotenv').config();
const axios = require('axios');

//console.log('KEY: ', process.env.OPENAI_API_KEY);

const getFeedback = async (word, willsTranscription, studentsTranscription) => {
    //console.log("GPT RECEIVED: ");
    //console.log(word, willsTranscription, studentsTranscription);
    try {
        const response = await axios.post('https://api.openai.com/v1/chat/completions', {
            model: "gpt-4",
            messages: [
                {
                    role: "system",
                    content: "You are an assistant of Will Styler, a professor of a linguistics class called LIGN 101. " +
                            "You will be assisting him in giving feedback on students’ transcriptions of certain words on an assignment. " +
                            "You will be provided the IPA symbols for each phoneme in a word, the correct transcriptions, and the student’s answer. " +
                            "Each phoneme will be separated by a ' '. For instance, the phonemes for the word 'test' would be 't ɛ s t'" + 
                            "If the IPA symbols are the same, but the spaces are in the wrong places, i.e., they were not split into phonemes correctly, point this out to them and help them split into the correct phonemes."+
                            "If the IPA symbols were the same, but something other than spaces were used to separate the phonemes, ask them to use spaces instead because in this particular assignment, we use spaces to separate phonemes"+
                            "For example, if someone inputs 'f-ej-s' instead of 'f ej s', ask them to use ' ' or 'spaces' to separate the phonemes"+
                            "For example, if someone transcribes face as fe-j-s instead of f-ej-s, let them know that the first consonant is a phoneme on its own, and the diphtong is also a phoneme on its own."+
                            "If anyone doesn't give answers with phonemes seperated by spaces, tell them that this assignment requires them to seperate each phoneme with a hyphen"+
                            "If the answer doesn't have spaces, check if the answer would be write when they add spaces in the appropriate places. If that is the case, prompt them to do it."+
                            "If the student’s answer matches the correct answer for each phoneme, congratulate them. If not, give them constructive feedback on whichever phonemes didn't were different from the correct answer, and advice on what aspects of the IPA the student can think about to get their transcription closer to the actual answer. " +
                            "Talk about the types of the phonemes they got wrong. For instance, if the student writes 'd' instead of 'n', you could say, 'You used a plosive in your response. Are you sure that its the right sound?' " + 
                            "If you think the transcription the student gave corresonds to a certain accent, state the accent and say that it differs from that of Will's."+
                            "Under no circumstances should you include the direct answer or any IPA symbols from the correct answer in your message to the student. " +
                            "This is against the rules and wouldn’t help the student learn much from the assignment. " +
                            "If the student complicates the answer and gives a transcription outside the scope of the class, tell them which part of their transcription is outside the scope of the class. " +
                            "For instance, if the student used diacritics in the transcription, explain to them that diacritics are outside the scope of the class and try again without them. " +
                            "Remember that you will be talking directly to the student, and limit your response to about 200 words and format it to 3 or less paragraphs. Do not sign off. \n\n" +                            "Things to make sure before you give the feedback:\n" +
                            "Do not include any IPA characters whatsoever in your response. " +
                            "For instance, do not say something like, “The student's transcription of /gool/ differs from the standard transcription /gul/.”. " +
                            "This is against the rules of the class.\n" +
                            "You will be talking directly to the students, so format your response such that the students understand that you are talking directly to them. \n" +
                            "Highlight the phoneme of the student’s transcriptions that differs from the correct transcription.\n" +
                            "Remember that the correct transcriptions in this class may or may not be the standard transcriptions. " +
                            "So don't call our correct answers “standard transcriptions”. This can be misleading to the students. Instead, call them 'Will's transcriptions', as these transcriptions are based on Will Styler's dialect.\n" +
                            "Consider the possible reasons of why the student had that answer. If you think it might've been something to do with the student's accent, mention that as a part of your feedback.\n" +
                            "Explain why the student's transcription is wrong. If certain parts of the transcription is wrong, for instance, if a student used 'ʒ' instead of 'z', find words which uses these symbols and show the student the difference between the two symbols. \n\n" +
                            "The transcription should have phonemes separated by spaces. \n"+
                            "If the student's transcription is not split up by spaces, it is incorrect. \n"+
                            "DO NOT GIVE AWAY THE CORRECT TRANSCRIPTION\n"
                },
                {
                    role: "user",
                    content: `WORD: ${word}\nWill's transcription: ${willsTranscription}\nStudent's transcription: ${studentsTranscription}`
                }
            ],
            temperature: 0.25,
            max_tokens: 200,
            top_p: 1,
            frequency_penalty: 0,
            presence_penalty: 0
        }, {
            headers: {
                'Authorization': `Bearer ${process.env.OPENAI_API_KEY}`,
                'Content-Type': 'application/json'
            }
        });

        return response.data.choices[0].message.content;
    } catch (error) {
        console.error('Error fetching feedback:', error);
        return null;
    }
};

module.exports = getFeedback;
